File: ru.kmorozov.api.onedrive/src/main/java/ru/kmorozov/onedrive/tasks/DeleteTask.java
package ru.kmorozov.onedrive.tasks;

import com.google.api.client.util.Preconditions;
import ru.kmorozov.onedrive.client.OneDriveItem;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.File;
import java.io.IOException;

public class DeleteTask extends Task {

    private static final Logger log = LogManager.getLogger(DeleteTask.class.getName());
    private final OneDriveItem remoteFile;
    private final File localFile;

    public DeleteTask(Task.TaskOptions options, OneDriveItem remoteFile) {

        super(options);

        this.remoteFile = Preconditions.checkNotNull(remoteFile);
        localFile = null;
    }

    public DeleteTask(Task.TaskOptions options, File localFile) {

        super(options);

        this.localFile = Preconditions.checkNotNull(localFile);
        remoteFile = null;
    }

    public int priority() {
        return 100;
    }

    @Override
    public String toString() {
        if (null != this.localFile) {
            return "Delete local file " + this.localFile.getPath();
        } else {
            return "Delete remote file " + this.remoteFile.getFullName();
        }
    }

    @Override
    protected void taskBody() throws IOException {
        if (null != this.localFile) {
            this.fileSystem.delete(this.localFile);
            this.reporter.localDeleted();
            DeleteTask.log.info("Deleted local file " + this.localFile.getPath());
        } else {
            this.api.delete(this.remoteFile);
            this.reporter.remoteDeleted();
            DeleteTask.log.info("Deleted remote file " + this.remoteFile.getFullName());
        }
    }
}

File: ru.kmorozov.api.onedrive/src/main/java/ru/kmorozov/onedrive/tasks/DeleteTask.kt
package ru.kmorozov.onedrive.tasks

import com.google.api.client.util.Preconditions
import org.apache.logging.log4j.LogManager
import ru.kmorozov.onedrive.client.OneDriveItem
import java.io.File
import java.io.IOException

class DeleteTask : Task {
    private val remoteFile: OneDriveItem?
    private val localFile: File?

    constructor(options: TaskOptions, remoteFile: OneDriveItem) : super(options) {

        this.remoteFile = Preconditions.checkNotNull(remoteFile)
        this.localFile = null
    }

    constructor(options: TaskOptions, localFile: File) : super(options) {

        this.localFile = Preconditions.checkNotNull(localFile)
        this.remoteFile = null
    }

    public override fun priority(): Int {
        return 100
    }

    override fun toString(): String {
        return if (null != localFile) {
            "Delete local file " + localFile.path
        } else {
            "Delete remote file " + remoteFile!!.fullName
        }
    }

    @Throws(IOException::class)
    public override fun taskBody() {
        if (null != localFile) {
            fileSystem.delete(localFile)
            reporter.localDeleted()
            log.info("Deleted local file " + localFile.path)
        } else {
            if (remoteFile != null) {
                api.delete(remoteFile)
                reporter.remoteDeleted()
                log.info("Deleted remote file " + remoteFile.fullName)
            }
        }
    }

    companion object {

        private val log = LogManager.getLogger(DeleteTask::class.java.name)
    }
}
